### PR TITLE
Don't allow non-formatting-context to be relayout-roots

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### 387e937dbc13f00ecad35bfe5800f81d0832ae09
<pre>
Don&apos;t allow non-formatting-context to be relayout-roots
<a href="https://bugs.webkit.org/show_bug.cgi?id=289140">https://bugs.webkit.org/show_bug.cgi?id=289140</a>
<a href="https://rdar.apple.com/146195256">rdar://146195256</a>

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium-review.googlesource.com/c/chromium/src/+/2707136">https://chromium-review.googlesource.com/c/chromium/src/+/2707136</a>

As lots of state can be passed from a non-fc to its sibling, treating
these as formatting context roots is incorrect for lots of edge cases.

This patch disallows non-fcs to be relayout-root.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::objectIsRelayoutBoundary):
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/normal-flow/margin-collapsing-dynamic-expected.xht:
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/normal-flow/margin-collapsing-dynamic.html:
</pre>
https://github.com/WebKit/WebKit/commit/387e937dbc13f00ecad35bfe5800f81d0832ae09

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94219 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99230 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96269 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14106 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22236 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71876 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29216 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10468 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85072 "3 api tests failed or timed out") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52221 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10158 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2757 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44064 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80390 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101275 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21271 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15492 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80882 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21523 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80264 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24810 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2176 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14469 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21255 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26434 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20942 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22683 "Built successfully") | | |

### CODE REVIEW

**Comment by Ahmad-S792** [comment]
Need to import WPT test and also run A/B test. Hence, this is in draft.

File: Source/WebCore/rendering/RenderObject.cpp [hunk 1]
```diff
@@ -563,6 +563,13 @@ static inline bool objectIsRelayoutBoundary(const RenderElement* object)
     if (object->style().width().isIntrinsicOrAuto() || object->style().height().isIntrinsicOrAuto() || object->style().height().isPercentOrCalculated() || object->style().width().isPercentOrCalculated())
         return false;
 
+    // A block which doesn't establish a new formating context can pass a whole
+    // bunch of state (floats, margins) to an arbitrary sibling, causing that
+    // sibling to position/size differently.
+    auto* renderBlock = dynamicDowncast<RenderBlock>(object);
+    if (!renderBlock->createsNewFormattingContext())
+        return false;
+
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderObject.cpp:572]: While it makes sense to stop dirty bit propagation at FC boundary (and not within) -and it also matches LFC design, this PR is slightly incomplete as we should also remove certain redundant checks.

**Comment by Ahmad-S792** [comment]
Fixed by Alan. Thanks @alanbaradlay.
